### ISSUE TITLE
Interaction custom augmentations and  COCOEvaluator

### ISSUE THREAD

**Issue by CyrilRJK** [facebookresearch/detectron2#3800]
I have a custom coco detectio  dataset with images of variable size. To ensure every example seen by the model is presented at the same resolution I apply the  T.FixedSizeCrop((SIZE, SIZE),  pad_value=0) augmentation to every image to take a random crop of images, and pad them if are smaller than a the specified SIZE. I wrote a custom trainer to do this (see code). In this trainer I also build the COCOEvaluator to automatically evaluate while training. To make sure the data is evaluated properly I apply the CROP transformation in both the train_data_loader and the test_data_loader but this does not seem to work as expected:

## Actual vs Expected behavior:

According to the COCOEvaluator the model cannot predict anything or only gets a few boxes right:

```
|  AP   |  AP50  |  AP75  |  APs  |  APm  |  APl  |
|:-----:|:------:|:------:|:-----:|:-----:|:-----:|
| 0.000 | 0.000  | 0.000  | 0.000 |  nan  |  nan  |
```

However, when I visualise the predictions they actually look quite good. And training for just 100 steps without the CROP actually produces COCOEvaluator results that make sense:
```
|   AP   |  AP50  |  AP75  |  APs   |  APm  |  APl  |
|:------:|:------:|:------:|:------:|:-----:|:-----:|
| 21.720 | 49.335 | 18.354 | 21.720 |  nan  |  nan  |
```

My question is, what am I missing about the interaction between the custom augmentations and the COCOEvaluator in Detectron2?

1. Full runnable code or full changes you made:
```
SIZE = 256
AUG = T.FixedSizeCrop((SIZE, SIZE),  pad_value=0)

class MyTrainer(DefaultTrainer):
    @classmethod
    def build_evaluator(cls, cfg, dataset_name, output_folder=None):
        if not output_folder:
            output_folder = os.path.join(cfg.OUTPUT_DIR, "validation")
        return COCOEvaluator(dataset_name=dataset_name, output_dir=output_folder, tasks=('bbox',)) 
    
    @classmethod
    def build_train_loader(cls, cfg):
        augmentations = [AUG]

        return build_detection_train_loader(cfg, 
                                            mapper=DatasetMapper(cfg, is_train=True, augmentations=augmentations),)
                                            # sampler=SAMPLER)
    
    @classmethod
    def build_test_loader(cls, cfg, name="test"):
        augmentations = [AUG]
        return build_detection_test_loader(cfg, name, mapper=DatasetMapper(cfg, is_train=False, augmentations=augmentations))

def train():
    cfg = get_cfg()
    
    """TEST"""
    cfg.TEST.AUG.ENABLED = True
        
    """model"""
    cfg.merge_from_file(
        model_zoo.get_config_file("COCO-Detection/faster_rcnn_X_101_32x8d_FPN_3x.yaml")
    )
    cfg.MODEL.ROI_HEADS.NUM_CLASSES = 1
    cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = (
        512
    )

    """datasets"""
    cfg.DATASETS.TRAIN = ("train",)
    cfg.DATASETS.TEST = ("test",)
    cfg.TEST.EVAL_PERIOD = 10000
    
    """dataloader"""
    cfg.DATALOADER.NUM_WORKERS = 1
    cfg.DATALOADER.FILTER_EMPTY_ANNOTATIONS = False

    """solver"""
    cfg.SOLVER.IMS_PER_BATCH = 8
    cfg.SOLVER.BASE_LR = 0.00001  # pick a good LR
    cfg.SOLVER.MAX_ITER = 50000
    cfg.SOLVER.STEPS = (1000, 10000, 20000) 
    cfg.SOLVER.WARMUP_ITERS = 10
    cfg.SOLVER.GAMMA = 0.5

    """misc"""
    cfg.SEED = 456
    cfg.OUTPUT_DIR = "/home/user/output/"
    os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
    
    """trainer"""
    trainer = MyTrainer(cfg)
    trainer.resume_or_load(resume=False)
    trainer.train()


train()
```

**Comment by Robotatron** [comment]
Did you find the issue? I have a similar issue with predictions working, but COCOEval showing NAN

**Comment by CyrilRJK** [comment]
> Did you find the issue? I have a similar issue with predictions working, but COCOEval showing NAN

Did not, the best solution if probably to write your own evaluation logic.

**Comment by jvdgoltz** [comment]
Hi, 
stumbled across this thread because we have a similar issue: Added augmentation and we are using the `RotatedCOCOEval` and it showed 0 mAP while training loss was perfectly converging.

I think it has to do with the fact that the ground truths for the evaluation are converted from the detectron dataset dicts to COCO format and cached as a json-file  __before__ any processing (resizing, cropping, augmentation, ...) happens. 

So you are now comparing model outputs on a cropped image with the ground truths on the unprocessed image, which is obviously a discrepancy and might be the reason for the bad mAP.

**Comment by Robotatron** [comment]
@jvdgoltz so the workaround would be to disable any TTA augs? Even disable resizing of the image? 

For my case a custom dataset A works fine, and a custom dataset B does not (COCO eval shows 100%, or 0% or NaN).

**Comment by jvdgoltz** [comment]
Yes. Either disable all image shape-changing test time augmentations or implement in the process() function of the evaluator that the model predictions are transformed back to the original image size.

**Comment by jvdgoltz** [comment]
Small correction to my previous post.
We introduced our bug by modifying the `height` and `width` field in the dataset dict in the mapper after augmentation, because the image has changed shape. However, that is not the correct way of using those fields. They indicate the original image size.
(See https://detectron2.readthedocs.io/en/latest/tutorials/models.html#model-input-format for the documentation of the input format)

In the `detectron2.modeling.postpressing.detector_postprocess()` function, the original image height and width are used to rescale the prediction to the original shape.

This means that cropping does not work correctly when doing inference, rescaling the image however is no problem.
